text/0000-offline-buildpackages.md

```diff
@@ -0,0 +1,455 @@
+- `asset-override` provides a way to override the `uri` used to download an asset.
+
+
+#### General Format:
+```toml
+[asset-cache (optional)]
+  name = "(required)"
+ 
+[[exclude-dependencies (optional)]]
+  id = "(required)"
+  version = "(required)"
+
+[[exclude-assets (optional)]]
+  sha256 = "(required)"
+
+[[asset-override (optional)]]
+  uri = "preferred/asset/location"
+  sha256 = "asset-sha256"
+```
+
+#### Example
+```toml
+[asset-cache]
+  name = "node-asset-cache"
+ 
+[[exclude-dependencies]]
+  id = "buildpacks/node-engine"
+  version = "1.2.3"
+
+[[exclude-assets]]
+  sha256 = "vulnerable-node10-sha256"
+  name = "Node js version with security vulnerability. (non-functional)"
+
+[[asset-override]]
+  uri = "../local/path/to/nodejs-asset"
+  sha256 = "node13-sha256"
+```
+
+
+Asset cache creation should:
+  - filter excluded assets specified in the `asset-config` file.
+  - Transform all remaining assets to an image layer filesystem changeset where the asset is provided at `/cnb/assets/<artifact-digest>`.
+  - Order all assets layers diffID.
+  - Add the `io.buildpacks.asset.layers` and `io.buildpacks.asset.metadata` label metadata to the asset image. If name is specified for `io.buildpacks.asset.metadata`, generate one.
+  - set the created time in image config to a constant value.
+  - set the modification time of all non-asset files to a constant value.
+
+#### UX in practice
+Likely the creation of asset packages will follow two different routes for implementation buildpacks and metabuildpackages. This dichotomy follows the existing differences between packaging metabuildpackages, and implementation buildpackages.
+
+##### Implementation Buildpacks
+
+running `pack package-asset <implementation-buildpack.toml>` should produce an asset cache, as a `buildpack.toml` file contains all needed `assets`.
+
+##### Metabuildpacks
+
+running `pack package-asset <metabuildpack-package.toml>` will use a `package.toml` file as this command needs to reference other `buildpackages` images in the `dependencies` array by specifying a `uri` or `image`.
+
+## Metabuildpackage Creation
+
+### Metadata
+
+When creating a metabuildpack, we want to retain all of the asset information from the child buildpacks.
+
+As such `assets` array in `io.buildpacks.buildpack.layers` `io.buildpacks.buildpack.metadata` of the resulting metabuildpack will be a sum of all of the `asset` arrays in its children.
+
+If there are conflicts when merging these entries, the platform may decide to continue or fail.
+
+
+## Builder Creation
+
+When creating a builder we can now specify `asset-cache` packages we wish to include in the builder image. We may also pass an `--asset-config asset.toml` file to excluded particular assets that would normally be included. Additionally there should be the option to include all assets from specified buildpackages using the `--offline` flag (the resulting builders will be very large).
```

Comment:
Would it make sense to basically inline the `asset.toml` schema under an `[assets]` table in `builder.toml`? And maybe also include the asset-cache URI as a builder.toml option. It might be nice for users if all the config for a given builder could be specified in one file, even if some things like `asset-cache` can be modified on the fly with a flag.